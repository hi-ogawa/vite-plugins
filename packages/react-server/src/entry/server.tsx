import { splitFirst } from "@hiogawa/utils";
import reactDomServer from "react-dom/server.edge";
import { injectRSCPayload } from "rsc-html-stream/server";
import { __global } from "../lib/global";
import {
  createModuleMap,
  initDomWebpackSsr,
  invalidateImportCacheOnFinish,
} from "../lib/ssr";
import { ENTRY_REACT_SERVER_WRAPPER, invalidateModule } from "../plugin/utils";

export async function handler(request: Request): Promise<Response> {
  const reactServer = await importReactServer();

  // server action and render rsc
  const result = await reactServer.handler({ request });
  if (result instanceof Response) {
    return result;
  }

  // ssr rsc
  const htmlStream = await renderHtml(result.stream);
  return new Response(htmlStream, {
    status: result.status,
    headers: {
      "content-type": "text/html",
    },
  });
}

export async function importReactServer(): Promise<
  typeof import("./react-server")
> {
  if (import.meta.env.DEV) {
    return __global.dev.reactServer.ssrLoadModule(
      ENTRY_REACT_SERVER_WRAPPER,
    ) as any;
  } else {
    return import("/dist/rsc/index.js" as string);
  }
}

export async function renderHtml(
  rscStream: ReadableStream,
): Promise<ReadableStream> {
  await initDomWebpackSsr();

  const { default: reactServerDomClient } = await import(
    "react-server-dom-webpack/client.edge"
  );

  const [rscStream1, rscStream2] = rscStream.tee();

  // use unique id for each render to simplify ssr module invalidation during dev
  // (see src/lib/ssr.tsx for details)
  const renderId = Math.random().toString(36).slice(2);

  const rscNode = await reactServerDomClient.createFromReadableStream(
    rscStream1,
    {
      ssrManifest: {
        moduleMap: createModuleMap({ renderId }),
        moduleLoading: null,
      },
    },
  );

  if (import.meta.env.DEV) {
    // ensure latest css
    invalidateModule(__global.dev.server, "\0virtual:ssr-assets");
    invalidateModule(__global.dev.server, "\0virtual:react-server-css.js");
    invalidateModule(__global.dev.server, "\0virtual:dev-ssr-css.css?direct");
  }
  const assets = (await import("virtual:ssr-assets" as string)).default;

  const ssrStream = await reactDomServer.renderToReadableStream(rscNode, {
    bootstrapModules: assets.bootstrapModules,
  });

  return ssrStream
    .pipeThrough(invalidateImportCacheOnFinish(renderId))
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(injectToHead(assets.head))
    .pipeThrough(new TextEncoderStream())
    .pipeThrough(injectRSCPayload(rscStream2));
}

function injectToHead(data: string) {
  const marker = "<head>";
  let done = false;
  return new TransformStream<string, string>({
    transform(chunk, controller) {
      if (!done && chunk.includes(marker)) {
        const [pre, post] = splitFirst(chunk, marker);
        controller.enqueue(pre + marker + data + post);
        done = true;
        return;
      }
      controller.enqueue(chunk);
    },
  });
}
