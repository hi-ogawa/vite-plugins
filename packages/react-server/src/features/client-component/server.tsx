import { tinyassert } from "@hiogawa/utils";
import ReactServer from "react-server-dom-webpack/server.edge";
import type { BundlerConfig, ImportManifestEntry } from "../../types/react";

// https://github.com/facebook/react/blob/c8a035036d0f257c514b3628e927dd9dd26e5a09/packages/react-server-dom-webpack/src/ReactFlightWebpackReferences.js#L43

// $$id: /src/components/counter.tsx#Counter
//   â‡•
// id: /src/components/counter.tsx
// name: Counter

/* @__NO_SIDE_EFFECTS__ */
export function registerClientReference(id: string, name: string) {
  // reuse everything but { $$async: true }.
  // `$$async` is not strictly necessary if we use `__webpack_chunk_load__` trick
  // but for now, we go with async `__webpack_require__` for simplicity.
  const reference = ReactServer.registerClientReference({}, id, name);
  return Object.defineProperties(
    {},
    {
      ...Object.getOwnPropertyDescriptors(reference),
      $$async: { value: true },
    },
  );
}

export function createBundlerConfig(): BundlerConfig {
  return new Proxy(
    {},
    {
      get(_target, $$id, _receiver) {
        tinyassert(typeof $$id === "string");
        let [id, name] = $$id.split("#");
        tinyassert(id);
        tinyassert(name);
        return { id, name, chunks: [] } satisfies ImportManifestEntry;
      },
    },
  );
}
