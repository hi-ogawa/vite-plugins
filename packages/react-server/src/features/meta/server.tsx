// @ts-ignore
import metadataFiles from "virtual:metadata-files";
import { objectHas } from "@hiogawa/utils";
import type { MetadataFilesExports } from "./plugin";
import { type Metadata, type Viewport, toViewportMetaContent } from "./utils";

export async function MetadataLinks({
  metadata,
  viewport,
}: {
  metadata: Metadata;
  viewport: Viewport;
}) {
  const title = getTitle(metadata);
  return (
    <>
      {(metadataFiles as MetadataFilesExports).favicon && (
        <link rel="icon" href="/favicon.ico" sizes="any" />
      )}
      <meta name="viewport" content={toViewportMetaContent(viewport)} />
      {typeof title === "string" && <title>{title}</title>}
      {typeof metadata.description === "string" && (
        <meta name="description" content={metadata.description} />
      )}
    </>
  );
}

function getTitle(m: Metadata): string | null {
  return typeof m.title === "string"
    ? m.title
    : objectHas(m.title, "default")
      ? m.title.default
      : null;
}
