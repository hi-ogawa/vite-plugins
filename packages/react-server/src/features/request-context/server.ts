import { createNextCookies } from "../next/cookie";
import type { RevalidationType } from "../server-component/utils";
import { createContextStorage } from "./utils";

const requestContextStorage = createContextStorage<RequestContext>();

export class RequestContext {
  nextCookies: ReturnType<typeof createNextCookies>;
  revalidate?: RevalidationType; // TODO: multiple revlidation paths
  private responseHeaders = new Headers();

  constructor(public requestHeaders: Headers) {
    this.nextCookies = createNextCookies(requestHeaders);
  }

  getResponseHeaders() {
    return {
      ...Object.fromEntries(this.responseHeaders.entries()),
      "set-cookie": this.nextCookies.toSetCookie(),
    };
  }

  mergeResponseHeaders(headers: Headers) {
    // TODO: not sure which should overwrite
    for (const [k, v] of headers) {
      this.responseHeaders.set(k, v);
    }
    this.nextCookies.mergeSetCookie(headers);
  }

  injectResponseHeaders(response: Response): Response {
    this.mergeResponseHeaders(response.headers);
    return new Response(response.body, {
      status: response.status,
      headers: this.getResponseHeaders(),
    });
  }

  run<T>(f: () => T): T {
    return requestContextStorage.run(this, f);
  }
}

function getRequestContext() {
  const context = requestContextStorage.getStore();
  if (!context) {
    // we tolerate non-existing context
    // since async storage is not well supported on stackblitz
    console.error("[WARNING] RequestContext not available");
    return new RequestContext(new Headers());
  }
  return context;
}

export function headers() {
  return getRequestContext().requestHeaders;
}

export function cookies() {
  return getRequestContext().nextCookies.cookies;
}

export function revalidatePath(path: string) {
  getRequestContext().revalidate = path;
}
