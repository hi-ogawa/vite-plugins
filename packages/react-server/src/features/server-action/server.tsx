import { memoize, tinyassert } from "@hiogawa/utils";
import ReactServer from "react-server-dom-webpack/server.edge";
import type { ReactServerErrorContext } from "../../server";
import type { BundlerConfig, ImportManifestEntry } from "../../types/react";

// https://github.com/facebook/react/blob/c8a035036d0f257c514b3628e927dd9dd26e5a09/packages/react-server-dom-webpack/src/ReactFlightWebpackReferences.js#L87

/* @__NO_SIDE_EFFECTS__ */
export function registerServerReference(
  action: Function,
  id: string,
  name: string,
) {
  if (typeof action !== "function") {
    return action;
  }
  return ReactServer.registerServerReference(action, id, name);
}

export type ActionResult = {
  error?: ReactServerErrorContext;
  data?: unknown;
  responseHeaders?: Record<string, string>;
};

const REFERENCE_SEP = "#";

export function createActionBundlerConfig(): BundlerConfig {
  return new Proxy(
    {},
    {
      get(_target, $$id, _receiver) {
        tinyassert(typeof $$id === "string");
        let [id, name] = $$id.split(REFERENCE_SEP);
        tinyassert(id);
        tinyassert(name);
        return {
          id,
          name,
          chunks: [],
        } satisfies ImportManifestEntry;
      },
    },
  );
}

// same as packages/react-server/src/features/use-client/server.tsx
export const serverReferenceImportPromiseCache = new Map<
  string,
  Promise<unknown>
>();

const serverReferenceWebpackRequire = memoize(importServerReference, {
  cache: serverReferenceImportPromiseCache,
});

export function initializeReactServer() {
  Object.assign(globalThis, {
    __vite_react_server_webpack_require__: serverReferenceWebpackRequire,
    __vite_react_server_webpack_chunk_load__: () => {
      throw new Error("todo: __webpack_chunk_load__");
    },
  });
}

async function importServerReference(id: string): Promise<unknown> {
  if (import.meta.env.DEV) {
    return await import(/* @vite-ignore */ id);
  } else {
    const mod = await import("virtual:server-references" as string);
    const dynImport = mod.default[id];
    tinyassert(dynImport, `server reference not found '${id}'`);
    return dynImport();
  }
}

export async function importServerAction(id: string): Promise<Function> {
  const [file, name] = id.split(REFERENCE_SEP) as [string, string];
  const mod: any = await importServerReference(file);
  return mod[name];
}
