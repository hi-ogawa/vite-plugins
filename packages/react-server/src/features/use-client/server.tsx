import { createDebug, memoize, tinyassert } from "@hiogawa/utils";
import type {
  ImportManifestEntry,
  ModuleMap,
  WebpackRequire,
} from "../../lib/types";

const debug = createDebug("react-server:ssr-import");

// __webpack_require__ is called at least (maybe exactly?) twice for preloadModule and requireModule
// https://github.com/facebook/react/blob/706d95f486fbdec35b771ea4aaf3e78feb907249/packages/react-server-dom-webpack/src/ReactFlightClientConfigBundlerWebpack.js
// We use `memoize` since `__webpack_require__` needs to return stable promise during each render.
// During dev, memoize cache is invalidated before each render,
// so the SSR will load fresh module in each time.
// Note that memoize cache invalidation doesn't need to be so precise at all
// since vite's ssrLoadModule has cache already
// and memoize cache is used only to make `ssrLoadModule`'s promise stable.
export const ssrImportPromiseCache = new Map<string, Promise<unknown>>();

const ssrWebpackRequire: WebpackRequire = memoize(ssrImport, {
  cache: ssrImportPromiseCache,
});

async function ssrImport(id: string) {
  debug("[__webpack_require__]", { id });
  if (import.meta.env.DEV) {
    // strip off `?t=` added for browser by noramlizeClientReferenceId
    id = id.split("?t=")[0]!;
    // transformed to "ssrLoadModule" during dev
    return import(/* @vite-ignore */ id);
  } else {
    const clientReferences = await import(
      "virtual:client-references" as string
    );
    const dynImport = clientReferences.default[id];
    tinyassert(dynImport, `client reference not found '${id}'`);
    return dynImport();
  }
}

export function initializeWebpackSsr() {
  Object.assign(globalThis, {
    __webpack_require__: ssrWebpackRequire,
    __webpack_chunk_load__: () => {
      throw new Error("todo: __webpack_chunk_load__");
    },
  });
}

export function createModuleMap(): ModuleMap {
  return new Proxy(
    {},
    {
      get(_target, id, _receiver) {
        return new Proxy(
          {},
          {
            get(_target, name, _receiver) {
              tinyassert(typeof id === "string");
              tinyassert(typeof name === "string");
              return {
                id,
                name,
                chunks: [],
              } satisfies ImportManifestEntry;
            },
          },
        );
      },
    },
  );
}
