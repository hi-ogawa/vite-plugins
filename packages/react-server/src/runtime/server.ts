export { registerClientReference } from "../features/client-component/server";
export { registerServerReference } from "../features/server-action/server";
