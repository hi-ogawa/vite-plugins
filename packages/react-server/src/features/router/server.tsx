import { sortBy } from "@hiogawa/utils";
import React from "react";
import { type ReactServerErrorContext, createError } from "../error/shared";
import { renderMetadata } from "../meta/server";
import type { Metadata } from "../meta/utils";
import type { RevalidationType } from "../server-component/utils";
import type { ApiRouteMoudle } from "./api-route";
import {
  type MatchNodeEntry,
  type TreeNode,
  createFsRouteTree,
  matchRouteTree,
  parseRoutePath,
  toMatchParamsObject,
  toRouteId,
} from "./tree";
import { LAYOUT_ROOT_NAME, isAncestorPath } from "./utils";

// cf. https://nextjs.org/docs/app/building-your-application/routing#file-conventions
export interface RouteModule {
  page?: {
    default: React.FC<PageProps>;
    metadata?: Metadata;
    generateStaticParams?: () => Promise<Record<string, string>[]>;
  };
  layout?: {
    default: React.FC<LayoutProps>;
    metadata?: Metadata;
  };
  error?: {
    // TODO: warn if no "use client"
    default: React.FC<ErrorPageProps>;
  };
  "not-found"?: {
    default: React.FC;
  };
  loading?: {
    default: React.FC;
  };
  template?: {
    default: React.FC<{ children?: React.ReactNode }>;
  };
  route?: ApiRouteMoudle;
}

export type RouteModuleKey = keyof RouteModule;

export type RouteModuleTree = TreeNode<RouteModule>;

export function generateRouteModuleTree(globEntries: Record<string, any>) {
  const { tree, entries } = createFsRouteTree<RouteModule>(globEntries);
  const manifest = getRouteModuleManifest(entries);
  return { tree, manifest };
}

// use own "use client" components as external
function importRuntimeClient(): Promise<typeof import("../../runtime/client")> {
  return import("@hiogawa/react-server/runtime/client" as string);
}

function renderPage(node: RouteModuleTree, props: PageProps) {
  const Page = node.value?.page?.default ?? ThrowNotFound;
  return <Page {...props} />;
}

async function renderLayout(
  node: RouteModuleTree,
  props: PageProps,
  { prefix, params }: MatchNodeEntry<RouteModule>,
) {
  const {
    ErrorBoundary,
    RedirectBoundary,
    NotFoundBoundary,
    RemountRoute,
    LayoutContent,
    LayoutMatchProvider,
  } = await importRuntimeClient();

  let acc = <LayoutContent name={prefix} />;
  acc = <RedirectBoundary>{acc}</RedirectBoundary>;

  const NotFoundPage = node.value?.["not-found"]?.default;
  if (NotFoundPage) {
    acc = (
      <NotFoundBoundary fallback={<NotFoundPage />}>{acc}</NotFoundBoundary>
    );
  }

  const ErrorPage = node.value?.error?.default;
  if (ErrorPage) {
    acc = <ErrorBoundary errorComponent={ErrorPage}>{acc}</ErrorBoundary>;
  }

  const LoadingPage = node.value?.loading?.default;
  if (LoadingPage) {
    acc = (
      <RemountRoute>
        <React.Suspense fallback={<LoadingPage />}>{acc}</React.Suspense>
      </RemountRoute>
    );
  }

  const TemplatePage = node.value?.template?.default;
  if (TemplatePage) {
    acc = (
      <RemountRoute>
        <TemplatePage>{acc}</TemplatePage>
      </RemountRoute>
    );
  }

  const Layout = node.value?.layout?.default;
  if (Layout) {
    acc = (
      <Layout key={prefix} {...props}>
        {acc}
      </Layout>
    );
  } else {
    acc = <React.Fragment key={prefix}>{acc}</React.Fragment>;
  }

  acc = <LayoutMatchProvider value={{ params }}>{acc}</LayoutMatchProvider>;
  return acc;
}

export async function renderRouteMap(
  tree: RouteModuleTree,
  request: Pick<Request, "url" | "headers">,
) {
  const url = new URL(request.url);
  const baseProps: Omit<BaseProps, "params"> = {
    url: serializeUrl(url),
    request: {
      url: request.url,
      headers: serializeHeaders(request.headers),
    },
    searchParams: Object.fromEntries(url.searchParams),
  };
  const metadata: Metadata = {};
  const layoutContentMap: Record<string, string> = {};
  const nodeMap: Record<string, React.ReactNode> = {};
  let parentLayout = LAYOUT_ROOT_NAME;
  const result = matchRouteTree(tree, url.pathname);
  for (const m of result.matches) {
    const routeId = toRouteId(m.prefix, m.type); // TODO: move to MatchNodeEntry
    layoutContentMap[parentLayout] = routeId;
    parentLayout = m.prefix;
    const props: BaseProps = {
      ...baseProps,
      params: toMatchParamsObject(m.params),
    };
    if (m.type === "layout") {
      nodeMap[routeId] = await renderLayout(m.node, props, m);
      Object.assign(metadata, m.node.value?.layout?.metadata);
    } else if (m.type === "page") {
      nodeMap[routeId] = renderPage(m.node, props);
      Object.assign(metadata, m.node.value?.page?.metadata);
    } else {
      m.type satisfies never;
    }
  }
  return {
    layoutContentMap,
    nodeMap,
    metadata: renderMetadata(metadata),
    params: result.params,
  };
}

export function getCachedRoutes(
  tree: RouteModuleTree,
  lastPathname: string,
  revalidations: (RevalidationType | undefined)[],
) {
  const routeIds: string[] = [];
  const { matches } = matchRouteTree(tree, lastPathname);
  for (const m of matches) {
    if (
      m.type === "layout" &&
      !revalidations.some((r) => r && isAncestorPath(r, m.prefix))
    ) {
      routeIds.push(toRouteId(m.prefix, m.type));
    }
  }
  return routeIds;
}

// TODO
// support SSR-ing not-found
// requires passing status without throwing
const ThrowNotFound: React.FC = () => {
  throw createError({ status: 404 });
};

type SerializedURL = {
  [k in keyof URL]: URL[k] extends string ? URL[k] : never;
};

function serializeUrl(url: URL): SerializedURL {
  const kv: any = {};
  for (const k in url) {
    const v = (url as any)[k];
    if (typeof v === "string") {
      kv[k] = v;
    }
  }
  return kv;
}

function serializeHeaders(headers: Headers): Record<string, string> {
  const kv: Record<string, string> = {};
  headers.forEach((v, k) => (kv[k] = v));
  return kv;
}

interface BaseProps {
  url: SerializedURL;
  request: {
    url: string;
    headers: Record<string, string>;
  };
  params: Record<string, string>;
  searchParams: Record<string, string>;
}

export interface PageProps extends BaseProps {}

export interface LayoutProps extends React.PropsWithChildren<BaseProps> {}

export interface ErrorPageProps {
  error: Error;
  serverError?: ReactServerErrorContext;
  reset: () => void;
}

type RouteModuleEntry = {
  pathname: string;
  module?: RouteModule;
  dynamic: boolean;
  format: (params: Record<string, string>) => string;
};

export type RouteModuleManifest = {
  entries: RouteModuleEntry[];
};

export function getRouteModuleManifest(
  entries: Record<string, RouteModule>,
): RouteModuleManifest {
  const result: RouteModuleManifest = { entries: [] };
  for (const [pathname, module] of Object.entries(entries)) {
    const { dynamic, format } = parseRoutePath(pathname);
    result.entries.push({
      pathname,
      module,
      dynamic,
      format,
    });
  }
  result.entries = sortBy(result.entries, (e) => e.pathname);
  return result;
}
