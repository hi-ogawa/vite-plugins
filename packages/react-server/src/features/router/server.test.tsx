import { describe, expect, it } from "vitest";
import { generateRouteModuleTree, renderRouteMap } from "./server";

describe(generateRouteModuleTree, () => {
  it("basic", async () => {
    const files = [
      "/layout.tsx",
      "/page.tsx",
      "/x/page.tsx",
      "/x/error.tsx",
      "/x/y/layout.tsx",
      "/x/y/page.tsx",
    ];
    const input = Object.fromEntries(
      files.map((k) => [k, async () => ({ default: k.toUpperCase() })]),
    );
    const { tree } = generateRouteModuleTree(input);
    expect(tree).toMatchSnapshot();

    async function testMatch(pathname: string) {
      const match = await renderRouteMap(tree, {
        url: "https://test.local" + pathname,
        headers: new Headers(),
      });
      return {
        // inject pathname for the ease of reading snapshot
        __pathname: pathname,
        ...match,
      };
    }

    expect(await testMatch("/x")).matchSnapshot();
    expect(await testMatch("/x/y")).matchSnapshot();
    expect(await testMatch("/z")).matchSnapshot();
  });
});
