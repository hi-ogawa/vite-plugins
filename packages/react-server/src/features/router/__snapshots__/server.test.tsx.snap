// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRouteModuleTree > basic 1`] = `
{
  "children": {
    "": {
      "children": {
        "x": {
          "children": {
            "y": {
              "value": {
                "layout": {
                  "default": "/X/Y/LAYOUT.TSX",
                },
                "page": {
                  "default": "/X/Y/PAGE.TSX",
                },
              },
            },
          },
          "value": {
            "error": {
              "default": "/X/ERROR.TSX",
            },
            "page": {
              "default": "/X/PAGE.TSX",
            },
          },
        },
      },
      "value": {
        "layout": {
          "default": "/LAYOUT.TSX",
        },
        "page": {
          "default": "/PAGE.TSX",
        },
      },
    },
  },
}
`;

exports[`generateRouteModuleTree > basic 2`] = `
{
  "__pathname": "/x",
  "layouts": {
    "/": </LAYOUT.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/x",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/x",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/x",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    >
      <RedirectBoundary>
        <LayoutContent
          name="/"
        />
      </RedirectBoundary>
    <//LAYOUT.TSX>,
    "/x": <React.Fragment>
      <ErrorBoundary
        errorComponent="/X/ERROR.TSX"
      >
        <RedirectBoundary>
          <LayoutContent
            name="/x"
          />
        </RedirectBoundary>
      </ErrorBoundary>
    </React.Fragment>,
  },
  "pages": {
    "/x": </X/PAGE.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/x",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/x",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/x",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    />,
  },
}
`;

exports[`generateRouteModuleTree > basic 3`] = `
{
  "__pathname": "/x/y",
  "layouts": {
    "/": </LAYOUT.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/x/y",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/x/y",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/x/y",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    >
      <RedirectBoundary>
        <LayoutContent
          name="/"
        />
      </RedirectBoundary>
    <//LAYOUT.TSX>,
    "/x": <React.Fragment>
      <ErrorBoundary
        errorComponent="/X/ERROR.TSX"
      >
        <RedirectBoundary>
          <LayoutContent
            name="/x"
          />
        </RedirectBoundary>
      </ErrorBoundary>
    </React.Fragment>,
    "/x/y": </X/Y/LAYOUT.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/x/y",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/x/y",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/x/y",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    >
      <RedirectBoundary>
        <LayoutContent
          name="/x/y"
        />
      </RedirectBoundary>
    <//X/Y/LAYOUT.TSX>,
  },
  "pages": {
    "/x/y": </X/Y/PAGE.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/x/y",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/x/y",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/x/y",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    />,
  },
}
`;

exports[`generateRouteModuleTree > basic 4`] = `
{
  "__pathname": "/z",
  "layouts": {
    "/": </LAYOUT.TSX
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/z",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/z",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/z",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    >
      <RedirectBoundary>
        <LayoutContent
          name="/"
        />
      </RedirectBoundary>
    <//LAYOUT.TSX>,
    "/z": <React.Fragment>
      <RedirectBoundary>
        <LayoutContent
          name="/z"
        />
      </RedirectBoundary>
    </React.Fragment>,
  },
  "pages": {
    "/z": <ThrowNotFound
      params={{}}
      request={
        {
          "headers": {},
          "url": "https://test.local/z",
        }
      }
      url={
        {
          "hash": "",
          "host": "test.local",
          "hostname": "test.local",
          "href": "https://test.local/z",
          "origin": "https://test.local",
          "password": "",
          "pathname": "/z",
          "port": "",
          "protocol": "https:",
          "search": "",
          "username": "",
        }
      }
    />,
  },
}
`;
